File: backends/node/services/auth/infra/src/repository/handlers/q/ping-db.ts
import type pg from "pg";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { PingDbInput as I, PingDbOutput as O, IPingDbHandler } from "@d2/auth-app";

export class PingDb extends BaseHandler<I, O> implements IPingDbHandler {
  private readonly pool: pg.Pool;

  constructor(pool: pg.Pool, context: IHandlerContext) {
    super(context);
    this.pool = pool;
  }

  protected async executeAsync(_input: I): Promise<D2Result<O | undefined>> {
    const start = performance.now();
    try {
      await this.pool.query("SELECT 1");
      const latencyMs = Math.round(performance.now() - start);
      return D2Result.ok({ data: { healthy: true, latencyMs } });
    } catch (err) {
      const latencyMs = Math.round(performance.now() - start);
      const error = err instanceof Error ? err.message : "Unknown error";
      return D2Result.ok({ data: { healthy: false, latencyMs, error } });
    }
  }
}

File: backends/dotnet/services/Geo/Geo.Infra/Repository/Handlers/R/PingDb.cs
// -----------------------------------------------------------------------
// <copyright file="PingDb.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Infra.Repository.Handlers.R;

using System.Diagnostics;
using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

using H = D2.Geo.App.Interfaces.Repository.Handlers.R.IRead.IPingDbHandler;
using I = D2.Geo.App.Interfaces.Repository.Handlers.R.IRead.PingDbInput;
using O = D2.Geo.App.Interfaces.Repository.Handlers.R.IRead.PingDbOutput;

/// <summary>
/// Handler for pinging the Geo database to verify connectivity.
/// </summary>
public partial class PingDb : BaseHandler<H, I, O>, H
{
    private readonly GeoDbContext r_db;

    /// <summary>
    /// Initializes a new instance of the <see cref="PingDb"/> class.
    /// </summary>
    ///
    /// <param name="db">
    /// The Geo database context.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public PingDb(
        GeoDbContext db,
        IHandlerContext context)
        : base(context)
    {
        r_db = db;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            await r_db.Database.ExecuteSqlRawAsync("SELECT 1", ct);
            sw.Stop();

            return D2Result<O?>.Ok(
                new O(true, sw.ElapsedMilliseconds, null));
        }
        catch (Exception ex)
        {
            sw.Stop();
            LogDatabasePingFailed(Context.Logger, ex, TraceId);

            return D2Result<O?>.Ok(
                new O(false, sw.ElapsedMilliseconds, ex.Message));
        }
    }

    /// <summary>
    /// Logs an error when the database ping fails.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "Database ping failed. TraceId: {TraceId}")]
    private static partial void LogDatabasePingFailed(ILogger logger, Exception exception, string? traceId);
}
